### PR TITLE
Refactor Subscriber/User Avatar Upload/Download

### PR DESCRIPTION
**Pull request by marrouchi**
# Motivation

In this PR: 
- Ensure upload directories exist in Attachment module's onApplicationBootStrap()
- Marke specific AttachmentService as deprecated in favor of using store() and download()
- Add rootDir param to the store and download to choose whether the avatar or uploads dir.
- Download Subscriber Avatar 
- Fix config by placing avatars under the upload Dir.
- Add a DB migration to Update old DBs : 
     ```
     1. Set the subscriber avatar attachment object id for subscriber.avatar instead of using foreign_id.
     2. Make sure the user avatars are moved to the "avatars" folder 
     ```
- Make the subscriber download avatar use id instead of foreign_id using the AttachmentService.download()
- Refactor all download methods to use AttachmentService.download()
- Make sure the PATCH /user/:userId/edit endpoint uploads any new avatar and updates the account
- Fix guard access to access own avatar and update own account.
- Update frontend accordingly to reflect the new API changes.

# How to test 

1. **Ensure Upload Directories Exist**:
   - Restart the application.
   - Verify that the required upload directories (e.g., `uploads` and `avatars`) are created during the application's bootstrap process.

2. **Deprecation of Specific `AttachmentService` Methods**:
   - Check any existing usage of deprecated `AttachmentService` methods with `store()` and `download()`.
   - Verify that these replacements work as intended without errors.

3. **Test `store()` and `download()` Methods with `rootDir` Parameter**:
   - Test the `store()` and `download()` methods for both `avatars` and `uploads` directories by passing the `rootDir` parameter.
   - Ensure files are uploaded to and downloaded from the correct directories.

4. **Subscriber Avatar Download**:
   - Verify the `download()` functionality for subscriber avatars, ensuring avatars are retrieved using the attachment ID, not the `foreign_id`.

5. **Database Migration**:
   - Apply the database migration script.
   - Verify DB:
     - Subscriber avatars are correctly updated in the database with attachment object IDs.
     - User avatars are moved to the `avatars` folder.
   - Check the database for consistency after migration.

6. **Refactored `download()` Method**:
   - Confirm all download operations now use the `AttachmentService.download()` method.
   - Verify the expected behavior for downloading attachments across the application.

7. **PATCH `/user/:userId/edit` Endpoint** - Test the endpoint to:
     - Upload a new avatar for a user.
     - Ensure the avatar is saved in the `avatars` directory.
     - Verify the user account is updated correctly in the database.

8. **Access Guards for Avatars and Account Updates** - Test access guards to:
     - Ensure users can access and download their own avatars.
     - Restrict access to downloading other users' avatars.
     - Allow users to update their own accounts but not others.

9. **Frontend Changes**:
   - Test all frontend changes that integrate with the new API endpoints and logic:
     - Uploading and updating avatars.
     - Downloading avatars for display in the UI.
   - Verify the frontend reflects the new API structure and handles errors gracefully.

10. **Regression Testing**:
    - Ensure no existing functionality is broken, including uploading/downloading attachments and avatar management.

11. **Edge Cases**:
    - Test edge cases such as:
      - Uploading non-image files as avatars or over-sized files.
      - Handling missing directories or files.
      - Updating and downloading avatars for users without previously existing avatars.

# Type of change:

Please delete options that are not relevant.

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# Checklist:

- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added unit tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: frontend/public/locales/fr/translation.json [hunk 1]
```diff
@@ -533,6 +533,7 @@
     "invite": "Inviter",
     "send": "Envoyer",
     "fields": "Champs",
+    "upload": "Télécharger",
```
Comment (yassinedorbozgithub) [inline frontend/public/locales/fr/translation.json:536]: I think that "télécharger" has a better meaning with the the download process not with the upload process.
I propose to use "Téléverser" as a french translation for the upload operation
image
https://support.google.com/googleplay/answer/11012086?hl=fr-CA&co=GENIE.Platform%3DDesktop

File: api/src/utils/helpers/fs.ts [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+  // Read the contents of the source folder
+  const files = await fs.promises.readdir(sourceFolder);
+
+  // Filter only files (skip directories)
+  const filePaths = await Promise.all(
+    files.map(async (file) => {
+      const filePath = join(sourceFolder, file);
+      const stat = await fs.promises.stat(filePath);
+      return stat.isFile() ? filePath : null;
+    }),
```
[the remark this answers is not in the document]
Comment (MohamedAliBouhaouala) [reply api/src/utils/helpers/fs.ts:56]: const filePaths = await files.reduce(async (accPromise, file) => {
    const acc = await accPromise; 
    const filePath = join(sourceFolder, file);
    const stat = await fs.promises.stat(filePath);
    if (stat.isFile()) {
      acc.push(filePath);
    }
    return acc;
  }, Promise.resolve([])); // Start with an empty array wrapped in a resolved promise
  
  
Avoid concurrent reading to control code execution and avoid potential performance overhead

File: api/src/utils/helpers/fs.ts [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+  );
+
+  // Move each file to the destination folder
```
Comment (MohamedAliBouhaouala) [inline api/src/utils/helpers/fs.ts:59]: Same remark as above  : 

const moveFiles = await filePaths.reduce<Promise<void>>(async (accPromise, filePath) => {
  await accPromise; // Ensure the previous promise is resolved before proceeding
  if (filePath !== null) {
    const fileName = basename(filePath);
    const destination = resolve(join(destinationFolder, fileName));
    await moveFile(filePath, destination, overwrite); // Perform the move operation sequentially
  }
  return Promise.resolve(); // Return a resolved promise for the next iteration
}, Promise.resolve());

return moveFiles

File: api/src/attachment/services/attachment.service.ts [hunk 4]
```diff
@@ -272,18 +278,24 @@ export class AttachmentService extends BaseService<Attachment> {
...
+    rootDir = config.parameters.uploadDir,
+  ): Promise<Buffer> {
     if (this.getStoragePlugin()) {
       return await this.getStoragePlugin().readAsBuffer(attachment);
     } else {
-      if (!fileExists(attachment.location)) {
+      const path = resolve(join(rootDir, attachment.location));
+
+      if (!fileExists(path)) {
         throw new NotFoundException('No file was found');
```
Comment (MohamedAliBouhaouala) [inline api/src/attachment/services/attachment.service.ts:295]: TODO: remove from the database an attachement that was physically deleted.

Comment (MohamedAliBouhaouala) [reply api/src/attachment/services/attachment.service.ts:295]: No worries. Just want to point it out cause I tested it as an edge case

File: frontend/public/locales/fr/translation.json (hunk 1 above)
Comment (marrouchi) [inline frontend/public/locales/fr/translation.json:536]:
```suggestion
    "upload": "Téléverser",
```

**Review by yassinedorbozgithub — APPROVED** [review]

**Review by abdou6666 — APPROVED** [review]

File: api/src/utils/helpers/fs.ts [hunk 5]
```diff
@@ -48,22 +48,19 @@ export async function moveFiles(
...
-  const movePromises = filePaths
-    .filter((filePath): filePath is string => filePath !== null)
-    .map((filePath) => {
-      const fileName = basename(filePath);
-      const destination = resolve(join(destinationFolder, fileName));
-      return moveFile(filePath, destination, overwrite);
-    });
-
-  await Promise.all(movePromises);
+  for (const filePath of filePaths) {
```
Comment (MohamedAliBouhaouala) [inline api/src/utils/helpers/fs.ts:61]: check filePaths is not empty before starting a new loop

Comment (marrouchi) [reply api/src/utils/helpers/fs.ts:61]: readdir() will either throw or return an empty array. For loop handles already a empty array and will not iterate.
